crates/turbopack-ecmascript/js/src/runtime.js

```diff
@@ -1,1004 +1,987 @@
+ * @param {ChunkPath} chunkPath
+ * @returns {boolean}
+ */
+function removeModuleFromChunk(moduleId, chunkPath) {
+  const moduleChunks = moduleChunksMap.get(moduleId);
+  moduleChunks.delete(chunkPath);
+
+  if (moduleChunks.size > 0) {
+    return false;
+  }
+
+  moduleChunksMap.delete(moduleId);
+  return true;
+}
+
+/**
+ * Instantiates a runtime module.
+ */
+/**
+ *
+ * @param {ModuleId} moduleId
+ * @returns {Module}
+ */
+function instantiateRuntimeModule(moduleId) {
+  return instantiateModule(moduleId, SourceType.Runtime);
+}
+
+/**
+ * Subscribes to chunk updates from the update server and applies them.
+ *
+ * @param {ChunkPath} chunkPath
+ */
+function subscribeToChunkUpdates(chunkPath) {
+  // This adds a chunk update listener once the handler code has been loaded
+  globalThis.TURBOPACK_CHUNK_UPDATE_LISTENERS.push([
+    chunkPath,
+    handleApply.bind(null, chunkPath),
+  ]);
+}
 
-  /** @type {Runtime} */
-  const runtime = {
-    loadedChunks,
-    modules: moduleFactories,
-    cache: moduleCache,
-    instantiateRuntimeModule,
-  };
+function markChunkAsLoaded(chunkPath) {
+  const chunkLoader = chunkLoaders.get(chunkPath);
+  if (!chunkLoader) {
+    loadedChunks.add(chunkPath);
 
-  /**
-   * @param {ChunkRegistration} chunkRegistration
-   */
-  function registerChunk([chunkPath, chunkModules, ...run]) {
-    markChunkAsLoaded(chunkPath);
-    subscribeToChunkUpdates(chunkPath);
-    for (const [moduleId, moduleFactory] of Object.entries(chunkModules)) {
-      if (!moduleFactories[moduleId]) {
-        moduleFactories[moduleId] = moduleFactory;
-      }
-      addModuleToChunk(moduleId, chunkPath);
-    }
-    runnable.push(...run);
-    runnable = runnable.filter((r) => r(runtime));
+    // This happens for all initial chunks that are loaded directly from
+    // the HTML.
+    return;
   }
 
-  self.TURBOPACK_CHUNK_UPDATE_LISTENERS =
-    self.TURBOPACK_CHUNK_UPDATE_LISTENERS || [];
-  self.TURBOPACK = { push: registerChunk };
-  chunksToRegister.forEach(registerChunk);
-})();
+  // Only chunks that are loaded via `loadChunk` will have a loader.
+  chunkLoader.onLoad();
+}
+
+/** @type {Runtime} */
+const runtime = {
+  loadedChunks,
+  modules: moduleFactories,
+  cache: moduleCache,
+  instantiateRuntimeModule,
+};
+
+/**
+ * @param {ChunkRegistration} chunkRegistration
+ */
+function registerChunk([chunkPath, chunkModules, ...run]) {
+  markChunkAsLoaded(chunkPath);
+  subscribeToChunkUpdates(chunkPath);
+  for (const [moduleId, moduleFactory] of Object.entries(chunkModules)) {
+    if (!moduleFactories[moduleId]) {
+      moduleFactories[moduleId] = moduleFactory;
+    }
+    addModuleToChunk(moduleId, chunkPath);
+  }
+  runnable.push(...run);
+  runnable = runnable.filter((r) => r(runtime));
+}
+
+globalThis.TURBOPACK_CHUNK_UPDATE_LISTENERS =
+  globalThis.TURBOPACK_CHUNK_UPDATE_LISTENERS || [];
+
+globalThis.TURBOPACK.forEach(registerChunk);
+globalThis.TURBOPACK = {
+  push: registerChunk,
+};
```

Comment:
This line causes `globalThis.TURBOPACK` to be inferred to be `{ push: ... }` by TypeScript, despite the explicit declaration in `index.d.ts`. I couldn't manage to fix this cleanly, any ideas?

Reply:
cc @ForsakenHarmony